app/controllers/live_streams_controller.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+class LiveStreamsController < ApplicationController
+  # Including ActionController::Live changes all actions in this controller!
+  # Therefore, it is extracted into a separate controller
+  include ActionController::Live
+
+  def download_submission_file
+    begin
+      @submission = authorize AuthenticatedUrlHelper.retrieve!(Submission, request, force_render_host: false)
+    rescue Pundit::NotAuthorizedError
+      # TODO: Option to disable?
+      # TODO: PR: privileged execution flag + Error / Exit Code
+      # Using the submission ID parameter would allow looking up the corresponding exercise ID
+      # Therefore, we just redirect to the root_path, but actually expect to redirect back (that should work!)
+      redirect_back(fallback_location: root_path, alert: t('exercises.download_file_tree.gone'))
+    end
+
+    desired_file = params[:filename].to_s
+    runner = Runner.for(current_user, @submission.exercise.execution_environment)
+    fallback_location = implement_exercise_path(@submission.exercise)
+    send_runner_file(runner, desired_file, fallback_location)
+  end
+
+  private
+
+  def send_runner_file(runner, desired_file, redirect_fallback = root_path)
+    filename = File.basename(desired_file)
+    send_stream(filename: filename, disposition: 'attachment') do |stream|
+      runner.download_file desired_file do |chunk, overall_size|
+        unless response.committed?
+          # Disable Rack::ETag, which would otherwise cause the response to be cached
+          # See https://github.com/rack/rack/issues/1619#issuecomment-848460528
+          response.set_header('Last-Modified', Time.now.httpdate)
+          response.set_header('Content-Length', overall_size)
+          # Commit the response headers immediately, as streaming would otherwise remove the Content-Length header
+          # This will prevent chunked transfer encoding from being used, which is okay as we know the overall size
+          # See https://github.com/rails/rails/issues/18714
+          response.commit!
+          stream.ignore_disconnect = true
```

Comment:
Fair point, I already thought about that when implementing. Unfortunately, I can only abort the data transfer by closing the underlying TCP connection; HTTP does not have a close mechanism I am aware of. In combination with the persistent keep-alive HTTP connections we use, this would also terminate all (other) pipelined requests.

Therefore, the only solution I can propose:
- Don't reuse an existing keep-alive connection when downloading files; always create a new connection (potentially adds some overhead for establishing the TCP connection and securing it with HTTPS).
- Use this new connection when transferring files
- Close the TCP connection when a user disconnects from CodeOcean.

Unfortunately, our request library Faraday in conjunction with the HTTP adapter does not support cancelling a request yet. Hence, we can either use another library, fix the library, or accept that we don't have the optimal solution for this edge-case. What do you think would make sense here?

When looking into the library locally (in order to find whether and how to add the cancellation feature), I noticed the following logs in Poseidon with a cancelled request. Are these errors expected there? The behaviour is reproducible with Postman cancelling a request as well:

```
time="2022-10-24T18:07:10.017129Z" level=error msg="Could not retrieve the requested file." error="communication with executor failed: nomad error during retrieve file content copy: error executing command in API: error executing command in allocation: context canceled" package=api
2022/10/24 18:07:10 http: superfluous response.WriteHeader call from github.com/openHPI/poseidon/pkg/logging.(*loggingResponseWriter).WriteHeader (logging.go:59)
time="2022-10-24T18:07:10.017201Z" level=debug msg="Influx data point" duration=4247004000 name=poseidon_/files/raw package=monitoring status=500
time="2022-10-24T18:07:10.017213Z" level=error msg="Failing /api/v1/runners/25-ef73bd3e-53ae-11ed-b82d-2e62a1fdfe45/files/raw" code=500 duration=4.247019s method=GET path=/api/v1/runners/25-ef73bd3e-53ae-11ed-b82d-2e62a1fdfe45/files/raw user_agent=PostmanRuntime/7.29.2
```

<hr>

I _tend_ to keep the current behaviour and still ignore the disconnect (maybe with an additional comment). Otherwise, I would fear that we invest too much time compared to the impact (which is unfortunate, I would really like to see this feature). What do you think?

Reply:
Okay, I was doing more research on this topic (and already had working patch for the library), but then identified that I might use a trick: As I am streaming the response, I can `break` in the lambda function used to handle the chunked bytes. If I do so, the outer context including the HTTP request is cancelled -- just as desired. I changed that.

This just leaves the 500 response generated by Poseidon - handing over to you for this one.

Reply:
That's great :tada: We catch the Poseidon error here: https://github.com/openHPI/poseidon/pull/239